dashboard/app/models/user_level.rb

```diff
@@ -108,6 +110,20 @@ def paired?
     driver? || navigator?
   end
 
+  def csf_validated?
+    Script.get_from_cache(script_id).csf? && Script.cache_find_level(level_id).validated?
+  end
+
+  # Record the time spent on a level by creating or updating the ValidatedUserLevel table
+  # for the UserLevel
+  def record_time_spent(time)
+    return unless time > 0 && csf_validated?
+    Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
+      vul = ValidatedUserLevel.find_or_create_by!(user_level_id: id)
+      ValidatedUserLevel.update(vul.id, time_spent: vul.time_spent + time)
+    end
```

Comment:
this will do two writes instead of one, in the case where we are creating the ValidatedUserLevel. Because this is a hot codepath, it seems worth optimizing. Here is one way we could eliminate the extra write:
```suggestion
    Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
      vul = ValidatedUserLevel.find_or_initialize_by(user_level_id: id)
      vul.time_spent += time
      vul.save!  
    end
```
because it is a hot codepath, please also add or update an integration test like https://github.com/code-dot-org/code-dot-org/blob/f3075a0d6888dfa2149748c7cbc942b78b3f04da/dashboard/test/integration/db_query_test.rb#L61 to supply a time in the milestone post, and verify the number of queries. It should sufficient to just test the case where the record does not exist.